Add links to resend renew email in back office

cintamani opened the pull request:
From: https://eaflood.atlassian.net/browse/RUBY-466

Add links in the BO dashboard and resource details page to resend a renewal email.
Add style of error and message flashes + partials.

<details> 
<summary> Screenshots: </summary>

<img width="974" alt="Screenshot 2019-08-13 at 16 48 31" src="https://user-images.githubusercontent.com/1385397/62957109-0d9f8380-bdec-11e9-855f-c50e2e66f736.png">
<img width="978" alt="Screenshot 2019-08-13 at 16 51 30" src="https://user-images.githubusercontent.com/1385397/62957110-0d9f8380-bdec-11e9-9d2f-2975fc861b8a.png">
<img width="983" alt="Screenshot 2019-08-13 at 16 59 56" src="https://user-images.githubusercontent.com/1385397/62957111-0d9f8380-bdec-11e9-90c3-d37c62c8e796.png">
<img width="976" alt="Screenshot 2019-08-13 at 17 00 32" src="https://user-images.githubusercontent.com/1385397/62957112-0d9f8380-bdec-11e9-8984-1415e8fd6b03.png">


</details>

cintamani committed:
Add links to resend renew email in back office

From: https://eaflood.atlassian.net/browse/RUBY-466

Add links in the BO dashboard and resource details page to resend a renewal email.
Add style of error and message flashes + partials.

cintamani committed:
Add messages in dashboard page

cintamani committed:
Add missing test coverage, fix bugs and kill bullet on error scenarios

cintamani committed:
Fix rubocop

cintamani commented on spec/requests/resend_renewal_email_spec.rb:
```diff
@@ -0,0 +1,85 @@
+        get request_path, {}, "HTTP_REFERER" => "/"
+        follow_redirect!
+
+        # binding.pry
```
```suggestion
```

cintamani committed:
Update spec/requests/resend_renewal_email_spec.rb

cintamani commented on spec/requests/resend_renewal_email_spec.rb:
```diff
@@ -0,0 +1,84 @@
+          example.run
+
+          Bullet.unused_eager_loading_enable = true
+        end
```
@irisfaraway ☝️ nice little trick to disable Bullet when he is unreasonable :D

cintamani committed:
Merge branch 'master' into 466-resend-renewal-email

cintamani commented on app/controllers/resend_renewal_email_controller.rb:
```diff
@@ -0,0 +1,33 @@
+    authorize
+
+    begin
+      RenewalReminderMailer.first_renew_with_magic_link_email(registration).deliver_now
```
```suggestion
      RenewalReminderMailer.first_reminder_email(registration).deliver_now
```

cintamani committed:
Update app/controllers/resend_renewal_email_controller.rb

cintamani commented on spec/requests/resend_renewal_email_spec.rb:
```diff
@@ -0,0 +1,84 @@
+
+      context "when an error happens", disable_bullet: true do
+        before do
+          expect(RenewalReminderMailer).to receive(:first_renew_with_magic_link_email).and_raise(StandardError)
```
```suggestion
          expect(RenewalReminderMailer).to receive(:first_reminder_email).and_raise(StandardError)
```

cintamani committed:
Update spec/requests/resend_renewal_email_spec.rb

Cruikshanks commented on app/assets/stylesheets/partials/_flash_messages.scss:
```diff
@@ -0,0 +1,16 @@
+.error-summary.wex_message {
```
It's going to have a knock on effect throughout the PR, but can we rename this as `flash_message` or something else more generic?

It is highly likely we will want to use this same decision and functionality in WCR renewals, so dropping the service specific prefix will make it easier to port.

cintamani replied:
Martin provided me with style and HTML for the messages. 
The main issue is that we are reusing style from the existing `error-summary` classes, which are defined in the Elements css, and give us responsiveness for free. This extra code here is just to adapt them to look as per screenshots. 
My suggestion is that we leave this as-it-is here and if this becomes an "important" bit later on before we move to the new GDS, then its worth spending more time tweaking this style.

Cruikshanks replied:
I'm fine with the styles, how it works etc. It's a great idea to reuse what we already have. I just don't want to use the prefix `wex` in this code.

Cruikshanks commented on app/controllers/resend_renewal_email_controller.rb:
```diff
@@ -0,0 +1,33 @@
+    begin
+      RenewalReminderMailer.first_reminder_email(registration).deliver_now
+
+      flash[:message] = I18n.t("resend_renewal_email.messages.success", email: registration.contact_email)
```
With this merged its going to make implementing [RUBY-467 Add option to resend confirmation email](https://eaflood.atlassian.net/browse/RUBY-467) much easier as well 🎉 😁 

So on the basis we know this other story is needed, and this is a pattern we'll want to use in WCR, is there a way we can move the implementation for assigning `flash[:message]`, and `flash[:error]` somewhere else? A controller concern for example?

Then rather than duplicating this across controllers, we can just call something like

```ruby
flash_success("Email sent. Well done you!")
flash_error("You broke it", "You appear to have tried using the system and broke it. Someone's in trouble!")

# ...

def flash_success(message)
  flash[:message]
end

def flash_error(message, description)
  flash[:error] = error
  flash[:error_details] = description
end
```

It is pre-emptive refactoring, but I'm 99.999% sure this will be reused!

cintamani replied:
Sure! Will do that :)

Cruikshanks reviewed (changes requested):
Left some specific comments. But essentially this PR not only adds the resend renewal link, it implements in-page confirmation messages for the first time in one of our back office apps.

I'm 100% certain it will be reused, not just in WEX but in WCR as well. As such I'm looking at this implementation with an eye to immediate reuse, hence my comments.

Ping me if you want to discuss this further. But other than that, this is awesome!

cintamani committed:
Add flash messages in concern

Cruikshanks committed:
Remove extra space in controller

We spotted there was a double space in the ResendRenewalEmailController.

Cruikshanks committed:
Remove wex prefix from styles

To simplify reuse if we need it this change replaces the `wex` prefix with something more generic.

Have tested locally and this change does not appear to have broken anything 😀

Cruikshanks committed:
Fix error in flash messages concern

Spotted `flash_error()` was referring to a variable that does not exist and when tested locally, was   causing an error, rather than just an error message to appear.

Cruikshanks committed:
Fix rubocop issues

Introduced by me removing the extra space. Forgot to re-align this stuff!

Cruikshanks commented:
@cintamani As we're all keen to get this into the next release I went ahead and committed the changes I was looking to see, plus a couple of fixes we spotted (extra space and the `flash_error()` error concern).

As you can see regards the CSS styles thing, I wasn't looking to refactor the code, simply swap the `wex` prefix for something more generic. Completely agree we should hold off investing too much time into it at this point until we have a firmer idea about implementing the design system.

Anyway, hope I've not stomped on your toes 😰 😬 !

irisfaraway reviewed (approved):
❤️ And flash messages too! Woohoo!

cintamani commented:
Not at all!! Glad you did it actually :) Thanks :)
